### ISSUE TITLE
Failed to specify default route for container

### ISSUE THREAD

**Issue by hickeng** [vmware/vic#2965]
Use of the port mapped port failing in build https://ci.vmware.run/vmware/vic/6434 ([logs here ](https://console.cloud.google.com/m/cloudstorage/b/vic-ci-logs/o/integration_logs_6434_4a9efa1111fcf94b15a4351482d985a37d22cbe9.zip?authuser=1  ))
Group: Group1-Docker-Commands.1-25-Docker-Port-Map
Failing container: 683564d24d9fdb512da8b369a037673a8cd3f6da8b9c955064453ef745b3046b

failing container
```
time="2016-11-01T04:51:07Z" level=info msg="setting ip address 172.16.0.3/16 for link eth0" 
time="2016-11-01T04:51:07Z" level=debug msg="added address 172.16.0.3/16 to link eth0" 
time="2016-11-01T04:51:07Z" level=debug msg="updateEndpoint(172.16.0.3/16, &{Common:{ExecutionEnvironment: ID:192 Name: Notes:} Static:false IP:172.16.0.3/16 Assigned:{IP:<nil> Mask:<nil>} Network:{Common:{ExecutionEnvironment: ID: Name:bridge Notes:} Type:bridge Gateway:{IP:172.16.0.1 Mask:ffff0000} Default:false Nameservers:[] Pools:[{FirstIP:172.16.0.0 LastIP:172.16.255.255}] Aliases:[]} DHCP:<nil>})" 
time="2016-11-01T04:51:07Z" level=debug msg="not setting route for network: default=false gateway=172.16.0.1" 
time="2016-11-01T04:51:07Z" level=debug msg="&{Common:{ExecutionEnvironment: ID:192 Name: Notes:} Static:false IP:172.16.0.3/16 Assigned:{IP:172.16.0.3 Mask:ffff0000} Network:{Common:{ExecutionEnvironment: ID: Name:bridge Notes:} Type:bridge Gateway:{IP:172.16.0.1 Mask:ffff0000} Default:false Nameservers:[] Pools:[{FirstIP:172.16.0.0 LastIP:172.16.255.255}] Aliases:[]} DHCP:<nil>}" 
```

passing container:
```
time="2016-11-01T14:49:58Z" level=info msg="setting ip address 172.16.0.3/16 for link eth0" ^M
time="2016-11-01T14:49:58Z" level=debug msg="added address 172.16.0.3/16 to link eth0" ^M
time="2016-11-01T14:49:58Z" level=debug msg="updateEndpoint(172.16.0.3/16, &{Common:{ExecutionEnvironment: ID:192 Name: Notes:} Static:false IP:172.16.0.3/16 Assigned:{IP:172.16.0.3 Mask:<nil>} Network:{Common:{ExecutionEnvironment: ID: Name:bridge Notes:} Type:bridge Gateway:{IP:172.16.0.1 Mask:ffff0000} Default:true Nameservers:[] Pools:[{FirstIP:172.16.0.0 LastIP:172.16.255.255}] Aliases:[]} DHCP:<nil>})" ^M
time="2016-11-01T14:49:58Z" level=info msg="Setting default gateway to 172.16.0.1" ^M
time="2016-11-01T14:49:58Z" level=info msg="updated default route to bridge interface, gateway: 172.16.0.1" ^M
```

Another item of concern for me here is that Static is showing as `false` when this is fundamentally a static IP configuration.

**Comment by hmahmood** [comment]
This does not seem to be reproducible locally or on the CI. The `Static=false` is expected; this is set to `true` only in the case when a container is created/run with `--ip`.

**Comment by hickeng** [comment]
Checking the vmware.log for the containerVM, the default route is explicitly set to false, rather than being omitted and that being the default boolean value:
```
 guestinfo.vice./networks|bridge/ip/IP = "AAAAAAAAAAAAAP//rBAAAw=="
 guestinfo.vice./networks|bridge/ip/Mask = "//8AAA=="
 guestinfo.vice..networks|bridge.network.gateway.Mask = "//8AAA=="
 guestinfo.vice..networks|bridge.network.gateway.IP = "AAAAAAAAAAAAAP//rBAAAQ=="
 guestinfo.vice./networks|bridge/Common/name = "<nil>"
 guestinfo.vice./networks|bridge/network/default = "false"
 guestinfo.vice./networks|bridge/network/pools|0/last = "rBD//w=="
 guestinfo.vice./networks|bridge/network/pools = "0"
 guestinfo.vice./networks|bridge/Common/ExecutionEnvironment = "<nil>"
  guestinfo.vice./networks = "bridge"
 guestinfo.vice./networks|bridge/static = "false"
 guestinfo.vice..networks|bridge.network.type = "bridge"
 guestinfo.vice./networks|bridge/network/pools|0/first = "rBAAAA=="
 guestinfo.vice./networks|bridge/network/Common/ExecutionEnvironment = "<nil>"
 guestinfo.vice./networks|bridge/network/Common/name = "bridge"
 networks|bridge/network/Common/notes = "<nil>"
 guestinfo.vice./networks|bridge/Common/id = "192"
 guestinfo.vice./networks|bridge/network/Common/id = "<nil>"
```

I do not see a path in context.go that would allow this, and it's a partial modification which negates possibility of having dropped the scope bind data completely.

**Comment by hmahmood** [comment]
@hickeng yes I couldn't see a path in context.go that would allow this, which is why this is so puzzling. Added to that we didn't see it in following CI runs, and both @andrewtchin and I could not reproduce this locally.

**Comment by hickeng** [comment]
I'm leaving it running as a background task locally - I've no idea how it could occur, but it did so will let it burn idle cpu for a while

**Comment by hickeng** [comment]
I cannot recreate locally, but it happened again: https://ci.vmware.run/vmware/vic/6483

```
time="2016-11-02T17:03:50Z" level=debug msg="&{Common:{ExecutionEnvironment: ID:192 Name: Notes:} Static:false IP:172.16.0.3/16 Assigned:{IP:<nil> Mask:<nil>} Network:{Common:{ExecutionEnvironment: ID: Name:bridge Notes:} Type:bridge Gateway:{IP:172.16.0.1 Mask:ffff0000} Default:false Nameservers:[]
```

**Comment by hickeng** [comment]
We explicitly preserve the ExtraConfig when refreshing handle state in commit.go as a workaround for callers of the portlayer not handling concurrent access errors.
However that means that we potentially discard changes made when we're racing with another operation that also modifies extraconfig.

Two operations get handles:
1. network.Bind
2. attach.Bind

network.Bind wins the race and updates the config first, setting default route and the IP address in a structure that wasn't present in the extraconfig previously (new keys)
attach.Bind updates with the VM hardware, but not the extraconfig, applying it's value of "false" to the default route, but not touching the IP because that structure wasn't present when they both acquired their handles.

**Comment by mlh78750** [comment]
Should be fixed with #3038
